File: scripts/scrape_research_images.py
"""
Fetch product/search page HTML and extract best product image URL.
Updates data/research/*.json with imageUrl fields.
"""
from __future__ import annotations

import json
import re
import time
import urllib.error
import urllib.request
from html import unescape
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
RESEARCH = ROOT / "data" / "research"

UA = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36"
)

OG_RE = re.compile(
    r'<meta[^>]+property=["\']og:image["\'][^>]+content=["\']([^"\']+)["\']',
    re.I,
)
OG_RE2 = re.compile(
    r'<meta[^>]+content=["\']([^"\']+)["\'][^>]+property=["\']og:image["\']',
    re.I,
)
# Amazon search result images
AMZ_IMG = re.compile(
    r'<img[^>]+class="[^"]*s-image[^"]*"[^>]+src="([^"]+)"',
    re.I,
)
AMZ_IMG2 = re.compile(
    r'src="(https://[^"]*media-amazon\.com/images/I/[^"]+)"',
    re.I,
)
# Jumia listing images
JUMIA_IMG = re.compile(
    r'data-src="(https://[^"]*jumia[^"]+\.(?:jpg|jpeg|png|webp)[^"]*)"',
    re.I,
)
JUMIA_IMG2 = re.compile(
    r'src="(https://[^"]*(?:jumia|jpmtuc)[^"]+\.(?:jpg|jpeg|png|webp)[^"]*)"',
    re.I,
)
LANDING = re.compile(
    r'data-old-hires="(https://[^"]+)"',
    re.I,
)


def fetch(url: str, timeout: int = 25) -> str:
    req = urllib.request.Request(
        url,
        headers={
            "User-Agent": UA,
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Accept-Language": "en-US,en;q=0.9,ar;q=0.8",
        },
        method="GET",
    )
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        raw = resp.read()
        charset = resp.headers.get_content_charset() or "utf-8"
        return raw.decode(charset, errors="replace")


def bad(url: str) -> bool:
    return bool(
        re.search(
            r"logo|icon|sprite|placeholder|badge|flag|avatar|pixel|1x1|spinner|transparent",
            url,
            re.I,
        )
    )


def normalize(url: str | None) -> str | None:
    if not url:
        return None
    url = unescape(url).strip()
    if url.startswith("//"):
        url = "https:" + url
    if bad(url) or url.startswith("data:"):
        return None
    # prefer larger jumia thumbs
    url = re.sub(r"/\d+x\d+/", "/680x680/", url)
    # amazon: strip size modifiers carefully — keep as-is if already good
    return url


def extract_image(html: str, page_url: str) -> str | None:
    # 1) og:image
    for rx in (OG_RE, OG_RE2):
        m = rx.search(html)
        if m:
            img = normalize(m.group(1))
            if img:
                return img

    # 2) amazon landing / product
    m = LANDING.search(html)
    if m:
        img = normalize(m.group(1))
        if img:
            return img

    # 3) amazon search images
    if "amazon." in page_url:
        m = AMZ_IMG.search(html)
        if m:
            img = normalize(m.group(1))
            if img:
                return img
        for m in AMZ_IMG2.finditer(html):
            img = normalize(m.group(1))
            if img and "._AC_" in img or "images/I/" in img:
                return img

    # 4) jumia
    if "jumia" in page_url:
        for rx in (JUMIA_IMG, JUMIA_IMG2):
            for m in rx.finditer(html):
                img = normalize(m.group(1))
                if img:
                    return img

    # 5) any large-ish product image host
    for m in re.finditer(
        r'(https://[^"\'\s]+(?:media-amazon|jumia|jpmtuc|ssl-images)[^"\'\s]+\.(?:jpg|jpeg|png|webp)[^"\'\s]*)',
        html,
        re.I,
    ):
        img = normalize(m.group(1))
        if img:
            return img

    return None


def main() -> None:
    stats = {"ok": 0, "fail": 0, "skip": 0}
    fails: list[str] = []

    for fp in sorted(RESEARCH.glob("*.json")):
        data = json.loads(fp.read_text(encoding="utf-8"))
        print(f"\n== {fp.name} ==")
        for item in data.get("items", []):
            iid = item.get("id", "?")
            url = item.get("url")
            if not url:
                stats["skip"] += 1
                print(f"  skip {iid} (no url)")
                continue
            try:
                html = fetch(url)
                img = extract_image(html, url)
                if img:
                    item["imageUrl"] = img
                    stats["ok"] += 1
                    print(f"  ok   {iid}")
                else:
                    stats["fail"] += 1
                    fails.append(f"{fp.name}:{iid}")
                    print(f"  fail {iid}")
            except Exception as e:
                stats["fail"] += 1
                fails.append(f"{fp.name}:{iid}:{type(e).__name__}")
                print(f"  err  {iid}: {type(e).__name__}: {e}")
            time.sleep(0.35)
        fp.write_text(json.dumps(data, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

    print("\nSummary:", stats)
    if fails:
        print("Failures:")
        for f in fails:
            print(" ", f)


if __name__ == "__main__":
    main()

File: scripts/scrape_research_images.js
/**
 * Scrape product images for research catalogs via Playwright.
 * Run: node scripts/scrape_research_images.js
 * Requires: playwright (npx playwright or local)
 */
const fs = require("fs");
const path = require("path");

const ROOT = path.join(__dirname, "..");
const RESEARCH = path.join(ROOT, "data", "research");

async function main() {
  let chromium;
  try {
    ({ chromium } = require("playwright"));
  } catch {
    console.error("Install playwright: npm i playwright");
    process.exit(1);
  }

  const files = fs.readdirSync(RESEARCH).filter((f) => f.endsWith(".json"));
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    userAgent:
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    locale: "en-US"
  });
  const page = await context.newPage();
  page.setDefaultTimeout(25000);

  for (const file of files) {
    const fp = path.join(RESEARCH, file);
    const data = JSON.parse(fs.readFileSync(fp, "utf8"));
    console.log("\n==", file, data.items.length, "items ==");
    for (const item of data.items) {
      if (item.imageUrl && !process.env.FORCE) {
        console.log("  skip (has image)", item.id);
        continue;
      }
      if (!item.url) {
        console.log("  skip (no url)", item.id);
        continue;
      }
      try {
        const img = await scrapeImage(page, item.url);
        if (img) {
          item.imageUrl = img;
          console.log("  ok", item.id, img.slice(0, 80));
        } else {
          console.log("  no img", item.id);
        }
      } catch (e) {
        console.log("  err", item.id, e.message.slice(0, 80));
      }
      // small delay
      await page.waitForTimeout(400);
    }
    fs.writeFileSync(fp, JSON.stringify(data, null, 2) + "\n", "utf8");
  }

  await browser.close();
  console.log("\nDone.");
}

async function scrapeImage(page, url) {
  await page.goto(url, { waitUntil: "domcontentloaded", timeout: 30000 });
  await page.waitForTimeout(1500);

  // Jumia product page
  if (url.includes("jumia.com.eg") && url.includes(".html") && !url.includes("/catalog/")) {
    const src = await page.evaluate(() => {
      const candidates = [
        ...document.querySelectorAll("img[data-src], img.src, img[src]"),
        ...document.querySelectorAll(".itm img, .-fh img, .sldr img, picture img")
      ];
      for (const img of candidates) {
        const s = img.getAttribute("data-src") || img.currentSrc || img.src || "";
        if (!s || s.startsWith("data:")) continue;
        if (/logo|icon|sprite|placeholder|badge|flag|avatar/i.test(s)) continue;
        if (/\.(jpg|jpeg|png|webp)/i.test(s) || s.includes("jumia")) return s.split("?")[0];
      }
      // og:image
      const og = document.querySelector('meta[property="og:image"]');
      if (og?.content) return og.content;
      return null;
    });
    return normalize(src);
  }

  // Jumia search
  if (url.includes("jumia.com.eg")) {
    const src = await page.evaluate(() => {
      const a = document.querySelector("a.core img");
      if (!a) return null;
      return a.getAttribute("data-src") || a.src || null;
    });
    return normalize(src);
  }

  // Amazon search or product
  if (url.includes("amazon.eg") || url.includes("amazon.com")) {
    const src = await page.evaluate(() => {
      const og = document.querySelector('meta[property="og:image"]');
      if (og?.content && !/sprite|//g.test(og.content) === false) {
        /* noop */
      }
      if (og?.content && !/sprite|icon/i.test(og.content)) return og.content;

      // product page main image
      const landing = document.querySelector("#landingImage, #imgTagWrapperId img, #main-image-container img");
      if (landing) {
        const s = landing.getAttribute("data-old-hires") || landing.currentSrc || landing.src;
        if (s && !s.startsWith("data:")) return s;
      }

      // search results
      const result = document.querySelector(
        '[data-component-type="s-search-result"] img.s-image, .s-main-slot img.s-image'
      );
      if (result) {
        const s = result.currentSrc || result.src;
        if (s && !s.startsWith("data:")) return s;
      }
      return null;
    });
    return normalize(src);
  }

  // generic og:image
  const src = await page.evaluate(() => {
    const og = document.querySelector('meta[property="og:image"]');
    return og?.content || null;
  });
  return normalize(src);
}

function normalize(src) {
  if (!src) return null;
  if (src.startsWith("//")) src = "https:" + src;
  // prefer larger jumia images if path has size tokens
  src = src.replace(/\/\d+x\d+\//, "/680x680/");
  return src;
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
